### PR TITLE
Allow endpoints with the same targetPort if they are on the same container component

### PR DESCRIPTION
**Pull request by amisevsk**
### What does this PR do?:
Implements the suggested solution in https://github.com/devfile/api/issues/821#issuecomment-1095411622, in particular:
* Endpoints using the same targetPort are allowed again, provided all endpoints using that targetPort are on the same container component
* Endpoints using the same targetPort on different containers return a validation error.

Documentation on endpoints and in validation-rule.md is updated to reflect this fact.

**Note:** this PR requires additional handling by tools that use endpoint names for containerPort names, as it's possible to generate invalid pod specs by not de-duplicating containerPorts -- cc: @kadel 

### Which issue(s) this PR fixes:
Closes https://github.com/devfile/api/issues/821

### PR acceptance criteria:
Testing and documentation do not need to be complete in order for this PR to be approved. We just need to ensure tracking issues are opened.

> - Open new test/doc issues under the [devfile/api](https://github.com/devfile/api/issues) repo
> - Check each criteria if:
>  - There is a separate tracking issue. Add the issue link under the criteria
>  **or**
>  - test/doc updates are made as part of this PR
> -  If unchecked, explain why it's not needed

- [x] Unit/Functional tests

  

- [ ] [QE Integration test](https://github.com/devfile/integration-tests) 

  

- [x] Documentation 

   

- [ ] Client Impact

### CODE REVIEW

**Review by elsony — CHANGES_REQUESTED** [review]

File: pkg/validation/validation-rule.md
Comment (amisevsk) [reply pkg/validation/validation-rule.md]: Embarrassing! Fixed, thank you :smile:

**Comment by elsony** [comment]
/lgtm 
Note: We should wait for @kadel to confirm if that is ok for odo before merging

File: pkg/validation/validation-rule.md
Comment (kadel) [inline pkg/validation/validation-rule.md:17]: If we allow the use of the same port on one container, why not allow the same ports on two different containers? Isn't basically the same situation as in most cases the containers will be in on Pod?


We don't need to block this PR on this.
We can revisit it in a separate issue.

**Review by kadel — APPROVED** [review]

File: pkg/validation/validation-rule.md
Comment (amisevsk) [reply pkg/validation/validation-rule.md:17]: Kubernetes won't allow multiple containerPorts within the same pod, so allowing the same port on different containers will result in the pod being rejected as invalid. For endpoints on the same container, only one containerPort is created (and only one service within the pod can listen on that port).

The endpoint defines more information than just the port. In particular, we use endpoints to define URL paths, protocols, and whether a route should be created. This results in multiple ingresses/routes that map to 1 or more services, that end up all pointing at one port in the container. An example would be something like
```
https://url1.com/path1 ───────────┐
                                  ▼
                           service port 8080 ──► container port 8080
                                  ▲
https://url2.com/path2 ───────────┘
```
Perhaps the ultimately better solution (requiring a new APIVersion) would be to extend endpoints to allow configuring something like this in one endpoint entry.

Comment (kadel) [reply pkg/validation/validation-rule.md:17]: > Kubernetes won't allow multiple containerPorts within the same pod, so allowing the same port on different containers will result in the pod being rejected as invalid. For endpoints on the same container, only one containerPort is created (and only one service within the pod can listen on that port).

oh, right.
In this case, we have quite a broken implementation of endpoints in devfile/library :-/
Currently, it maps every endpoint into the container port.
By re-introducing this change devfile/library will have to start doing smarter things with endpoints than just blindly mapping 1:1 to container ports.
/cc @elsony @yangcao77 

I think that this is another proof of how badly we need a common layer that controls the basic principles of how Devfile is translated to k8s resources. 
If we can't have it on code level in for of some shared library then at least we need to have documentation describing the ground rules of how devfile is translated to k8s resources. Otherwise, problems like this will keep coming.

**Review by elsony — APPROVED** [review]

**Review by ibuziuk — APPROVED** [review]

File: pkg/validation/validation-rule.md
Comment (yangcao77) [reply pkg/validation/validation-rule.md:17]: Yes. I have aware that library need to make change accordingly. But library can still use the endpoint name, if multiple endpoints uses same port, library can use the first endpoint uses that port. Or Odo want to change back to the old behavior? all use `port-<portNumber>` instead?

**Pull request merged**
